text/1009-group-by.rst

```diff
@@ -0,0 +1,394 @@
+              {name: 'Sprite', cost: 1},
+              {name: 'Giant eagle', cost: 2},
+              {name: 'Djinn', cost: 4}
+          ],
+          grouping: ['element']
+      }
+  ]
+
+
+Computing the average cost of each "element" that a card can have::
+
+  SELECT (GROUP Card BY .element) {
+      element := .key.element,
+      avg_cost := sum(.elements.cost) / count(.elements),
+  } ORDER BY .element
+
+::
+
+   [
+       {el: 'Air', avg_cost: 2.3333333333333335},
+       {el: 'Earth', avg_cost: 2.0},
+       {el: 'Fire', avg_cost: 3.0},
+       {el: 'Water', avg_cost: 2.5}
+   ]
+
+Computing the ratio of each card's cost to the average of its element::
+
+  SELECT (
+    FOR g in (GROUP Card BY .element) UNION (
+      WITH U := g.elements,
+      SELECT U {
+          name,
+          cost_ratio := .cost / math::mean(g.elements.cost)
+      })
+  ) ORDER BY .name;
+
+::
+
+  [
+      {name: 'Imp', cost_ratio: 0.3333333333333333},
+      {name: 'Dragon', cost_ratio: 1.6666666666666667},
+      {name: 'Bog monster', cost_ratio: 0.8},
+      {name: 'Giant turtle', cost_ratio: 1.2},
+      {name: 'Dwarf', cost_ratio: 0.5},
+      {name: 'Golem', cost_ratio: 1.5},
+      {name: 'Sprite', cost_ratio: 0.42857142857142855},
+      {name: 'Giant eagle', cost_ratio: 0.8571428571428571},
+      {name: 'Djinn', cost_ratio: 1.7142857142857142}
+  ]
+
+Counting the number of cards in each possible "element", "number of
+owners" combination bucket, as well as those things individually::
+
+  SELECT (
+    GROUP Card BY .element, nowners := count(.owners)
+    USING CUBE (element, nowners)
+  ) {
+      key: {element, nowners},
+      num := count(.elements),
+      grouping
+  } ORDER BY .grouping THEN .key.element THEN .key.nowners;
+
+::
+
+   [
+       {key: {element: [], nowners: []}, num: 9, grouping: []},
+       {key: {element: 'Air', nowners: []}, num: 3, grouping: ['element']},
+       {key: {element: 'Earth', nowners: []}, num: 2, grouping: ['element']},
+       {key: {element: 'Fire', nowners: []}, num: 2, grouping: ['element']},
+       {key: {element: 'Water', nowners: []}, num: 2, grouping: ['element']},
+       {key: {element: [], nowners: 1}, num: 1, grouping: ['nowners']},
+       {key: {element: [], nowners: 2}, num: 5, grouping: ['nowners']},
+       {key: {element: [], nowners: 3}, num: 1, grouping: ['nowners']},
+       {key: {element: [], nowners: 4}, num: 2, grouping: ['nowners']},
+       {key: {element: 'Air', nowners: 2}, num: 3, grouping: ['nowners', 'element']},
+       {key: {element: 'Earth', nowners: 2}, num: 1, grouping: ['nowners', 'element']},
+       {key: {element: 'Earth', nowners: 3}, num: 1, grouping: ['nowners', 'element']},
+       {key: {element: 'Fire', nowners: 1}, num: 1, grouping: ['nowners', 'element']},
+       {key: {element: 'Fire', nowners: 2}, num: 1, grouping: ['nowners', 'element']},
+       {key: {element: 'Water', nowners: 4}, num: 2, grouping: ['nowners', 'element']}
+   ]
```

Comment:
We need examples that involve FOR.

Reply:
See "Computing the ratio of each card's cost to the average of its element"